include/fir/Type.h

```diff
@@ -0,0 +1,280 @@
+struct FIRReferenceTypeStorage;
+struct FIRSequenceTypeStorage;
+struct FIRTypeDescTypeStorage;
+}
+
+enum FIRTypeKind {
+  // The enum starts at the range reserved for this dialect.
+  FIR_TYPE = mlir::Type::FIRST_FIR_TYPE,
+  FIR_BOX,
+  FIR_BOXCHAR,
+  FIR_BOXPROC,
+  FIR_CHARACTER,  // intrinsic
+  FIR_COMPLEX,  // intrinsic
+  FIR_DERIVED,  // derived
+  FIR_DIMS,
+  FIR_FIELD,
+  FIR_HEAP,
+  FIR_INT,  // intrinsic
+  FIR_LOGICAL,  // intrinsic
+  FIR_POINTER,
+  FIR_REAL,  // intrinsic
+  FIR_REFERENCE,
+  FIR_SEQUENCE,
+  FIR_TYPEDESC,
+};
+
+template<class... Ts> struct Typecase : Ts... { using Ts::operator()...; };
+template<class... Ts> Typecase(Ts...)->Typecase<Ts...>;
+
+bool isa_fir_type(mlir::Type);
+bool isa_std_type(mlir::Type t);
+bool isa_fir_or_std_type(mlir::Type t);
+
+// Intrinsic types
+
+class CharacterType : public mlir::Type::TypeBase<CharacterType, mlir::Type,
+                          detail::FIRCharacterTypeStorage> {
+public:
+  using Base::Base;
+  static CharacterType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_CHARACTER;
+  }
+
+  int getSizeInBits() const;
+  KindTy getFKind() const;
+};
+
+class IntType : public mlir::Type::TypeBase<IntType, mlir::Type,
+                        detail::FIRIntTypeStorage> {
+public:
+  using Base::Base;
+  static IntType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_INT; }
+
+  int getSizeInBits() const;
+  KindTy getFKind() const;
+};
+
+class LogicalType : public mlir::Type::TypeBase<LogicalType, mlir::Type,
+                        detail::FIRLogicalTypeStorage> {
+public:
+  using Base::Base;
+  static LogicalType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_LOGICAL; }
+
+  int getSizeInBits() const;
+  KindTy getFKind() const;
+};
+
+class RealType : public mlir::Type::TypeBase<RealType, mlir::Type,
+                     detail::FIRRealTypeStorage> {
+public:
+  using Base::Base;
+  static RealType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_REAL; }
+
+  int getSizeInBits() const;
+  KindTy getFKind() const;
+};
+
+class CplxType : public mlir::Type::TypeBase<CplxType, mlir::Type,
+                        detail::FIRCplxTypeStorage> {
+public:
+  using Base::Base;
+  static CplxType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_COMPLEX; }
+
+  int getSizeInBits() const;
+  KindTy getFKind() const;
+};
+
+// FIR support types
+
+class BoxType : public mlir::Type::TypeBase<BoxType, mlir::Type,
+                    detail::FIRBoxTypeStorage> {
+public:
+  using Base::Base;
+  static BoxType get(mlir::Type eleTy);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_BOX; }
+  mlir::Type getEleTy() const;
```

Comment:
Is `Ele` a common abbreviation in MLIR/LLVM ? If not, I think `getElementType` would be more readable and not that much longer.

Reply:
What's most common in mlir?  Given a choice, we prefer to spell out most names.  There are exceptions, Statement, for example.

Reply:
I am not going to change the name.